<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Michael Love" />


<title>EDA, QC, and DESeq2 / edgeR analysis of 109 RNA-seq samples of immunotherapy treatment</title>

<script src="preNivolumabOnNivolumab_files/header-attrs-2.11/header-attrs.js"></script>
<script src="preNivolumabOnNivolumab_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="preNivolumabOnNivolumab_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="preNivolumabOnNivolumab_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="preNivolumabOnNivolumab_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="preNivolumabOnNivolumab_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="preNivolumabOnNivolumab_files/navigation-1.1/tabsets.js"></script>
<link href="preNivolumabOnNivolumab_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="preNivolumabOnNivolumab_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">EDA, QC, and DESeq2 / edgeR analysis of 109 RNA-seq samples of immunotherapy treatment</h1>
<h4 class="author">Michael Love</h4>

</div>


<pre class="r"><code>x &lt;- read.csv(&quot;GSE91061_BMS038109Sample.hg19KnownGene.raw.csv.gz&quot;, row.names=1)
condition &lt;- factor(sub(&quot;.+_(.+)_.+&quot;, &quot;\\1&quot;, colnames(x)))
table(condition)</code></pre>
<pre><code>## condition
##  On Pre 
##  58  51</code></pre>
<pre class="r"><code>library(DESeq2)
dds &lt;- DESeqDataSetFromMatrix(x, 
                              colData=data.frame(condition), 
                              ~condition)
vsd &lt;- vst(dds, blind=FALSE)
plotPCA(vsd)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pca1-1.png" width="672" /></p>
<pre class="r"><code>rv &lt;- rowVars(assay(vsd))
pc &lt;- prcomp(t(assay(vsd)[head(order(-rv),1000),]))
plot(pc$x[,1:2], col=condition)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pcaForOutlier-1.png" width="672" /></p>
<pre class="r"><code>idx &lt;- pc$x[,1] &lt; -25
sum(idx)</code></pre>
<pre><code>## [1] 9</code></pre>
<pre class="r"><code>plot(pc$x[,1:2], col=idx+1, pch=20, asp=1)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pcaForOutlier-2.png" width="672" /></p>
<pre class="r"><code>condition &lt;- condition[!idx]
dds &lt;- dds[,!idx]</code></pre>
<p>For comparison, use minimal filtering with edgeR.</p>
<pre class="r"><code>library(edgeR)
y &lt;- DGEList(counts=counts(dds), group=condition)
keep &lt;- filterByExpr(y)
table(keep)</code></pre>
<pre><code>## keep
## FALSE  TRUE 
##  5956 16231</code></pre>
<pre class="r"><code>y &lt;- y[keep,]
dds &lt;- dds[keep,]</code></pre>
<p>We can see there is still structure in the 2D PCA, which is not related to the known covariate <code>condition</code>.</p>
<pre class="r"><code>vsd &lt;- vst(dds, blind=FALSE)
plotPCA(vsd)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pca2-1.png" width="672" /></p>
<p>Run simple DESeq2 analysis comparing the two groups, without attempting to control for the technical variation:</p>
<pre class="r"><code>system.time({
  dds &lt;- DESeq(dds, test=&quot;LRT&quot;, reduced=~1, fitType=&quot;glmGamPoi&quot;)
})</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## using &#39;glmGamPoi&#39; as fitType. If used in published research, please cite:
##     Ahlmann-Eltze, C., Huber, W. (2020) glmGamPoi: Fitting Gamma-Poisson
##     Generalized Linear Models on Single Cell Count Data. Bioinformatics.
##     https://doi.org/10.1093/bioinformatics/btaa1009</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>## Fit reduced model</code></pre>
<pre><code>## Calculate quasi likelihood ratio</code></pre>
<pre><code>## Prepare results</code></pre>
<pre><code>##    user  system elapsed 
##  11.694   1.730  14.602</code></pre>
<pre class="r"><code>res &lt;- results(dds)
table(res$padj &lt; .1)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 15392   839</code></pre>
<p>Ignoring the technical variation is not appropriate, if there is correlation between the technical variation and the condition, not including variables in the design formula that control for the effect on the expression estimates will lead to invalid inference, regardless of the method we choose.</p>
<p>We can estimate the technical variation with a number of methods, including RUV, SVA, or PEER. Here we demonstrate usage of RUV:</p>
<pre class="r"><code>library(RUVSeq)
set &lt;- newSeqExpressionSet(counts(dds))
set &lt;- betweenLaneNormalization(set, which=&quot;upper&quot;)
not_sig &lt;- rownames(res)[which(res$pvalue &gt; .1)]
empirical &lt;- rownames(set)[ rownames(set) %in% not_sig ]
set &lt;- RUVg(set, empirical, k=5)</code></pre>
<p>The factors of unwanted variation are labelled <code>W_1</code>, <code>W_2</code>, etc. and are stored as metadata. We add the original condition variable:</p>
<pre class="r"><code>pdat &lt;- pData(set)
pdat$condition &lt;- condition</code></pre>
<p>We can visualize how the factors of unwanted variation describe the samples in the PC1 and PC2 space:</p>
<pre class="r"><code>vsd$W1 &lt;- pdat$W_1
vsd$W2 &lt;- pdat$W_2
plotPCA(vsd, intgroup=&quot;W1&quot;)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pca3-1.png" width="672" /></p>
<pre class="r"><code>plotPCA(vsd, intgroup=&quot;W2&quot;)</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/pca3-2.png" width="672" /></p>
<p>Adding the factors to the design, and performing a LRT. Here we use <code>glmGamPoi</code> which is an efficient method for estimating dispersion when we have many samples.</p>
<pre class="r"><code>colData(dds) &lt;- cbind(colData(dds), pdat[,1:5])
design(dds) &lt;- ~W_1 + W_2 + W_3 + W_4 + W_5 + condition</code></pre>
<pre class="r"><code>system.time({
  dds &lt;- DESeq(dds, test=&quot;LRT&quot;, reduced=~W_1 + W_2 + W_3 + W_4 + W_5, 
               fitType=&quot;glmGamPoi&quot;)
})</code></pre>
<pre><code>## using pre-existing size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## found already estimated dispersions, replacing these</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## using &#39;glmGamPoi&#39; as fitType. If used in published research, please cite:
##     Ahlmann-Eltze, C., Huber, W. (2020) glmGamPoi: Fitting Gamma-Poisson
##     Generalized Linear Models on Single Cell Count Data. Bioinformatics.
##     https://doi.org/10.1093/bioinformatics/btaa1009</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>## Fit reduced model</code></pre>
<pre><code>## Calculate quasi likelihood ratio</code></pre>
<pre><code>## Prepare results</code></pre>
<pre><code>##    user  system elapsed 
##  23.558   2.016  27.222</code></pre>
<p>Controlling for technical variation in this case reduces the number of DE genes. It could also be the opposite, that controlling for technical variation increased the apparent number of DE genes. That the number is reduced here indicates that some of the previous results were likely due to confounding of technical variation with the condition variable. Ignoring that confounding, again, will result in invalid inference for all methods.</p>
<pre class="r"><code>res &lt;- results(dds)
table(res$padj &lt; .1)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 16212    19</code></pre>
<pre class="r"><code>res_sig &lt;- res[which(res$padj &lt; .1),]</code></pre>
<pre class="r"><code>plotMA(res, ylim=c(-5,5))</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/maplot-1.png" width="672" /></p>
<pre class="r"><code>y &lt;- calcNormFactors(y)
design &lt;- model.matrix(~W_1 + W_2 + W_3 + W_4 + W_5 + condition, data=pdat)
y &lt;- estimateDisp(y, design)
qlfit &lt;- glmQLFit(y, design)
qlft &lt;- glmQLFTest(qlfit)</code></pre>
<pre class="r"><code>tt &lt;- topTags(qlft, n=nrow(y))[[1]]
sum(tt$FDR &lt; .1)</code></pre>
<pre><code>## [1] 0</code></pre>
<pre class="r"><code>hist(tt$F, freq=FALSE)
F &lt;- tt[rownames(res_sig),&quot;F&quot;]
lines(density(F[!is.na(F)]))</code></pre>
<p><img src="preNivolumabOnNivolumab_files/figure-html/edgerDESeq2Compare-1.png" width="672" /></p>
<pre class="r"><code>match(rownames(res_sig), rownames(tt))</code></pre>
<pre><code>##  [1]  8 12  7  3  2 15 14 13  4  1 11 18  6 19  5 16  9 21 17</code></pre>
<pre class="r"><code>table(rownames(res_sig) %in% head(rownames(tt), 20))</code></pre>
<pre><code>## 
## FALSE  TRUE 
##     1    18</code></pre>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] RUVSeq_1.28.0               EDASeq_2.28.0               ShortRead_1.52.0           
##  [4] GenomicAlignments_1.30.0    Rsamtools_2.10.0            Biostrings_2.62.0          
##  [7] XVector_0.34.0              BiocParallel_1.28.3         edgeR_3.36.0               
## [10] limma_3.50.1                DESeq2_1.34.0               SummarizedExperiment_1.24.0
## [13] Biobase_2.54.0              MatrixGenerics_1.6.0        matrixStats_0.61.0         
## [16] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1         IRanges_2.28.0             
## [19] S4Vectors_0.32.3            BiocGenerics_0.40.0         testthat_3.1.2             
## [22] rmarkdown_2.11              devtools_2.4.3              usethis_2.1.5              
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-3          rjson_0.2.21              hwriter_1.3.2            
##   [4] ellipsis_0.3.2            rprojroot_2.0.2           fs_1.5.2                 
##   [7] farver_2.1.0              remotes_2.4.2             bit64_4.0.5              
##  [10] AnnotationDbi_1.56.2      fansi_1.0.2               xml2_1.3.3               
##  [13] codetools_0.2-18          R.methodsS3_1.8.1         splines_4.1.2            
##  [16] sparseMatrixStats_1.6.0   cachem_1.0.6              geneplotter_1.72.0       
##  [19] knitr_1.37                pkgload_1.2.4             jsonlite_1.8.0           
##  [22] annotate_1.72.0           dbplyr_2.1.1              R.oo_1.24.0              
##  [25] png_0.1-7                 compiler_4.1.2            httr_1.4.2               
##  [28] assertthat_0.2.1          Matrix_1.4-0              fastmap_1.1.0            
##  [31] cli_3.2.0                 htmltools_0.5.2           prettyunits_1.1.1        
##  [34] tools_4.1.2               gtable_0.3.0              glue_1.6.2               
##  [37] GenomeInfoDbData_1.2.7    dplyr_1.0.8               rappdirs_0.3.3           
##  [40] Rcpp_1.0.8                jquerylib_0.1.4           vctrs_0.3.8              
##  [43] rtracklayer_1.54.0        DelayedMatrixStats_1.16.0 xfun_0.29                
##  [46] stringr_1.4.0             ps_1.6.0                  brio_1.1.3               
##  [49] lifecycle_1.0.1           restfulr_0.0.13           XML_3.99-0.9             
##  [52] MASS_7.3-55               zlibbioc_1.40.0           scales_1.1.1             
##  [55] aroma.light_3.24.0        hms_1.1.1                 parallel_4.1.2           
##  [58] RColorBrewer_1.1-2        curl_4.3.2                yaml_2.3.5               
##  [61] memoise_2.0.1             ggplot2_3.3.5.9000        sass_0.4.0               
##  [64] biomaRt_2.50.3            latticeExtra_0.6-29       stringi_1.7.6            
##  [67] RSQLite_2.2.10            highr_0.9                 genefilter_1.76.0        
##  [70] BiocIO_1.4.0              desc_1.4.0                GenomicFeatures_1.46.5   
##  [73] filelock_1.0.2            pkgbuild_1.3.1            rlang_1.0.1              
##  [76] pkgconfig_2.0.3           bitops_1.0-7              evaluate_0.15            
##  [79] lattice_0.20-45           glmGamPoi_1.6.0           purrr_0.3.4              
##  [82] labeling_0.4.2            bit_4.0.4                 processx_3.5.2           
##  [85] tidyselect_1.1.2          magrittr_2.0.2            R6_2.5.1                 
##  [88] generics_0.1.2            DelayedArray_0.20.0       DBI_1.1.2                
##  [91] pillar_1.7.0              withr_2.4.3               survival_3.2-13          
##  [94] KEGGREST_1.34.0           RCurl_1.98-1.6            tibble_3.1.6             
##  [97] crayon_1.5.0              utf8_1.2.2                BiocFileCache_2.2.1      
## [100] progress_1.2.2            jpeg_0.1-9                locfit_1.5-9.4           
## [103] grid_4.1.2                blob_1.2.2                callr_3.7.0              
## [106] digest_0.6.29             xtable_1.8-4              R.utils_2.11.0           
## [109] munsell_0.5.0             bslib_0.3.1               sessioninfo_1.2.2</code></pre>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
